File: VB/WebDAVServer.FileSystemStorage.AspNet/DavContext.vb
Imports System
Imports System.IO
Imports System.Linq
Imports System.Net
Imports System.Runtime.InteropServices
Imports System.Security.Principal
Imports System.Web
Imports System.Configuration
Imports System.Threading.Tasks
Imports ITHit.Server
Imports ITHit.WebDAV.Server
Imports ITHit.WebDAV.Server.Acl
Imports ITHit.WebDAV.Server.Quota
Imports ILogger = ITHit.Server.ILogger
Imports WebDAVServer.FileSystemStorage.AspNet.ExtendedAttributes

''' <summary>
''' Implementation of <see cref="ContextAsync{IHierarchyItem}"/> .
''' Resolves hierarchy items by paths.
''' </summary>
Public Class DavContext
    Inherits ContextWebAsync(Of IHierarchyItem)

    ''' <summary>
    ''' Path to the folder which become available via WebDAV.
    ''' </summary>
    Public Property RepositoryPath As String

    ''' <summary>
    ''' Gets WebDAV Logger instance.
    ''' </summary>
    Public Property Logger As ILogger

    ''' <summary>
    ''' Singleton instance of <see cref="WebSocketsService"/> .
    ''' </summary>
    Public Property socketService As WebSocketsService
        Get
            Return WebSocketsService.Service
        End Get

        Private Set(ByVal value As WebSocketsService)
        End Set
    End Property

    ''' <summary>
    ''' Initializes a new instance of the DavContext class.
    ''' </summary>
    ''' <param name="httpContext"><see cref="HttpContext"/>  instance.</param>
    Public Sub New(httpContext As HttpContext)
        MyBase.New(httpContext)
        Logger = WebDAVServer.FileSystemStorage.AspNet.Logger.Instance
        RepositoryPath = If(ConfigurationManager.AppSettings("RepositoryPath"), String.Empty)
        Dim isRoot As Boolean = New DirectoryInfo(RepositoryPath).Parent Is Nothing
        Dim configRepositoryPath As String = If(isRoot, RepositoryPath, RepositoryPath.TrimEnd(Path.DirectorySeparatorChar))
        RepositoryPath = If(configRepositoryPath.StartsWith("~"), HttpContext.Current.Server.MapPath(configRepositoryPath), configRepositoryPath)
        Dim attrStoragePath As String =(If(ConfigurationManager.AppSettings("AttrStoragePath"), String.Empty)).TrimEnd(Path.DirectorySeparatorChar)
        attrStoragePath = If(attrStoragePath.StartsWith("~"), HttpContext.Current.Server.MapPath(attrStoragePath), attrStoragePath)
        If Not FileSystemInfoExtension.IsUsingFileSystemAttribute Then
            If Not String.IsNullOrEmpty(attrStoragePath) Then
                FileSystemInfoExtension.UseFileSystemAttribute(New FileSystemExtendedAttribute(attrStoragePath, Me.RepositoryPath))
            ElseIf Not(New DirectoryInfo(RepositoryPath).IsExtendedAttributesSupported()) Then
                Dim tempPath = Path.Combine(Path.GetTempPath(), System.Reflection.Assembly.GetExecutingAssembly().GetName().Name)
                FileSystemInfoExtension.UseFileSystemAttribute(New FileSystemExtendedAttribute(tempPath, Me.RepositoryPath))
            End If
        End If

        If Not Directory.Exists(RepositoryPath) Then
            WebDAVServer.FileSystemStorage.AspNet.Logger.Instance.LogError("Repository path specified in Web.config is invalid.", Nothing)
        End If
    End Sub

    ''' <summary>
    ''' Creates <see cref="IHierarchyItem"/>  instance by path.
    ''' </summary>
    ''' <param name="path">Item relative path including query string.</param>
    ''' <returns>Instance of corresponding <see cref="IHierarchyItem"/>  or null if item is not found.</returns>
    Public Overrides Async Function GetHierarchyItemAsync(path As String) As Task(Of IHierarchyItem)
        path = path.Trim({" "c, "/"c})
        'remove query string.
        Dim ind As Integer = path.IndexOf("?"c)
        If ind > -1 Then
            path = path.Remove(ind)
        End If

        Dim item As IHierarchyItem = Nothing
        item = Await DavFolder.GetFolderAsync(Me, path)
        If item IsNot Nothing Then Return item
        item = Await DavFile.GetFileAsync(Me, path)
        If item IsNot Nothing Then Return item
        Logger.LogDebug("Could not find item that corresponds to path: " & path)
        Return Nothing
    End Function

    ''' <summary>
    ''' Returns the physical file path that corresponds to the specified virtual path on the Web server.
    ''' </summary>
    ''' <param name="relativePath">Path relative to WebDAV root folder.</param>
    ''' <returns>Corresponding path in file system.</returns>
    Friend Function MapPath(relativePath As String) As String
        'Convert to local file system path by decoding every part, reversing slashes and appending
        'to repository root.
        Dim encodedParts As String() = relativePath.Split({"/"}, StringSplitOptions.RemoveEmptyEntries)
        Dim decodedParts As String() = encodedParts.Select(Of String)(AddressOf EncodeUtil.DecodeUrlPart).ToArray()
        Return Path.Combine(RepositoryPath, String.Join(Path.DirectorySeparatorChar.ToString(), decodedParts))
    End Function
End Class

File: CS/WebDAVServer.FileSystemStorage.AspNet/DavContext.cs
using System;
using System.IO;
using System.Linq;
using System.Net;
using System.Runtime.InteropServices;
using System.Security.Principal;
using System.Web;
using System.Configuration;
using System.Threading.Tasks;

using ITHit.Server;
using ITHit.WebDAV.Server;
using ITHit.WebDAV.Server.Acl;
using ITHit.WebDAV.Server.Quota;
using ILogger = ITHit.Server.ILogger;
using WebDAVServer.FileSystemStorage.AspNet.ExtendedAttributes;



namespace WebDAVServer.FileSystemStorage.AspNet
{
    /// <summary>
    /// Implementation of <see cref="ContextAsync{IHierarchyItemAsync}"/>.
    /// Resolves hierarchy items by paths.
    /// </summary>
    public class DavContext :
        ContextWebAsync<IHierarchyItemAsync>
    {

        /// <summary>
        /// Path to the folder which become available via WebDAV.
        /// </summary>
        public string RepositoryPath { get; private set; }

        /// <summary>
        /// Gets WebDAV Logger instance.
        /// </summary>
        public ILogger Logger { get; private set; }

        /// <summary>
        /// Singleton instance of <see cref="WebSocketsService"/>.
        /// </summary>
        public WebSocketsService socketService { get { return WebSocketsService.Service; } private set { } }

        /// <summary>
        /// Initializes a new instance of the DavContext class.
        /// </summary>
        /// <param name="httpContext"><see cref="HttpContext"/> instance.</param>
        public DavContext(HttpContext httpContext) : base(httpContext)
        {
            Logger = WebDAVServer.FileSystemStorage.AspNet.Logger.Instance;
            RepositoryPath = ConfigurationManager.AppSettings["RepositoryPath"] ?? string.Empty;
            bool isRoot = new DirectoryInfo(RepositoryPath).Parent == null;
            string configRepositoryPath = isRoot ? RepositoryPath : RepositoryPath.TrimEnd(Path.DirectorySeparatorChar);
            RepositoryPath = configRepositoryPath.StartsWith("~") ?
                HttpContext.Current.Server.MapPath(configRepositoryPath) : configRepositoryPath;

            string attrStoragePath = (ConfigurationManager.AppSettings["AttrStoragePath"] ?? string.Empty).TrimEnd(Path.DirectorySeparatorChar);
            attrStoragePath = attrStoragePath.StartsWith("~") ?
                HttpContext.Current.Server.MapPath(attrStoragePath) : attrStoragePath;

            if (!FileSystemInfoExtension.IsUsingFileSystemAttribute)
            {
                if (!string.IsNullOrEmpty(attrStoragePath))
                {
                    FileSystemInfoExtension.UseFileSystemAttribute(new FileSystemExtendedAttribute(attrStoragePath, this.RepositoryPath));
                } 
                else if (!(new DirectoryInfo(RepositoryPath).IsExtendedAttributesSupported()))
                {
                    var tempPath = Path.Combine(Path.GetTempPath(), System.Reflection.Assembly.GetExecutingAssembly().GetName().Name);
                    FileSystemInfoExtension.UseFileSystemAttribute(new FileSystemExtendedAttribute(tempPath, this.RepositoryPath));
                }
            }


            if (!Directory.Exists(RepositoryPath))
            {
                WebDAVServer.FileSystemStorage.AspNet.Logger.Instance.LogError("Repository path specified in Web.config is invalid.", null);
            }
        }

        /// <summary>
        /// Creates <see cref="IHierarchyItemAsync"/> instance by path.
        /// </summary>
        /// <param name="path">Item relative path including query string.</param>
        /// <returns>Instance of corresponding <see cref="IHierarchyItemAsync"/> or null if item is not found.</returns>
        public override async Task<IHierarchyItemAsync> GetHierarchyItemAsync(string path)
        {
            path = path.Trim(new[] { ' ', '/' });

            //remove query string.
            int ind = path.IndexOf('?');
            if (ind > -1)
            {
                path = path.Remove(ind);
            }

            IHierarchyItemAsync item = null;

            item = await DavFolder.GetFolderAsync(this, path);
            if (item != null)
                return item;

            item = await DavFile.GetFileAsync(this, path);
            if (item != null)
                return item;

            Logger.LogDebug("Could not find item that corresponds to path: " + path);

            return null; // no hierarchy item that corresponds to path parameter was found in the repository
        }

        /// <summary>
        /// Returns the physical file path that corresponds to the specified virtual path on the Web server.
        /// </summary>
        /// <param name="relativePath">Path relative to WebDAV root folder.</param>
        /// <returns>Corresponding path in file system.</returns>
        internal string MapPath(string relativePath)
        {
            //Convert to local file system path by decoding every part, reversing slashes and appending
            //to repository root.
            string[] encodedParts = relativePath.Split(new[] { "/" }, StringSplitOptions.RemoveEmptyEntries);
            string[] decodedParts = encodedParts.Select<string, string>(EncodeUtil.DecodeUrlPart).ToArray();
            return Path.Combine(RepositoryPath, string.Join(Path.DirectorySeparatorChar.ToString(), decodedParts));
        }
    }
}
